### ISSUE TITLE
Table dimensions are read as "0 by 0" by Windows Narrator / `role="textbox"` considered invalid in `td` and `th`

### ISSUE THREAD

**Issue by lkszzajac** [ckeditor/ckeditor5#16923]
## 📝 Provide detailed reproduction steps (if any)

1. Open any editor and remove its initial content (e.g. https://ckeditor.com/ckeditor-5/demo/feature-rich/)
2. Create a table (e.g. 3x3, can be empty)
3. Enable Windows Narrator
4. Focus the table widget

### ✔️ Expected result

Narrator reads the widget as a table and announces its dimensions

### ❌ Actual result

Narrator reads the widget as a table and reports its dimensions as 0x0

https://github.com/user-attachments/assets/fb1ce81a-e0de-45df-9f26-0a3d6630dc1c

## 📃 Other details

* Browser: Edge
* OS: Windows 11
* First affected CKEditor version: …
* Installed CKEditor plugins: …
* Related issues in other screen readers:
    * https://github.com/ckeditor/ckeditor5/issues/11867
    * https://github.com/ckeditor/ckeditor5/issues/12224
    * https://github.com/ckeditor/ckeditor5/issues/13479
* Issue does not reproduce in other screen readers because no other reader reports on table dimension by default on focus. VoiceOver can read the table dimensions only after focusing a cell and going "up" in focus by one level

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by lkszzajac** [comment]
We've checked the behavior in these setups:

Windows/Edge:
- Narrator - announces the table with "Enter table, 0x0"
- JAWS - announces the table with "Edit, land, type of text"
- NVDA - announces the table with "Figure, table, mutliline, blank"

macOS/Chrome:
- VoiceOver - announces the table with "table" without any dimensions.

macOS/Firefox/Safari:
- VoiceOver - announces the table with dimensions correctly, does not announce the cells as editable

**Comment by Comandeer** [comment]
According to the ARIA spec, elements with the `row` role (and [`<tr>` elements have it by default](https://w3c.github.io/html-aria/#:~:text=or%20aria%2D*%20attributes-,tr,role%3Drow,-If%20the%20ancestor)) can only contain [children with the `cell` role](https://w3c.github.io/aria/#row:~:text=Accessibility%20Child%20Roles%3A-,cell,rowheader,-Supported%20States%20and). However, due to how we implemented nested editables, we overwrite the role of the `<td>` elements to `textbox`, creating an accessibility tree that is invalid according to the specification. Probably we should rethink our approach to nested editables in tables.

Some possible solutions:

*   **remove the `textbox` role** – however it could lead to loss of information about cells being editable; needs testing,
*   **move the `textbox` role into a dynamically injected `div` element that would reside inside the cell** – sounds quite brittle and error-prone,
*   **keep the current structure and provide the table info ourselves** – that's a really dirty hack, though.

**Comment by Mati365** [comment]
I checked few solutions:

> remove the textbox role – however it could lead to loss of information about cells being editable; needs testing,

It works, but it stopped announcing that content is editable. At this moment, after focusing the first cell of the table, it's announcing:  `Tabela, przejdź do tabeli, zero na zero, edytuj, spacja` but after modification: `Tabela, Przejdź do tabeli, 5 na 6, wejście angielski, kolumna 3 z 5, koniec wiersza`. So imo, it's even worse. 

> move the textbox role into a dynamically injected div element that would reside inside the cell – sounds quite brittle and error-prone,

I tried to add `role: textbox` to `ck-table-bogus-paragraph` that is directly under `td` and t's not announcing info about editable content, identically as with the missing `role: textbox`.

> keep the current structure and provide the table info ourselves – that's a really dirty hack, though.

I'm not sure. Do you mean adding `aria-label` to `td`?

**Comment by Comandeer** [comment]
What screen reader did you test? VoiceOver? It would be better to test it in English and to also test JAWS and NVDA.

> It works, but it stopped announcing that content is editable. At this moment, after focusing the first cell of the table, it's announcing: `Tabela, przejdź do tabeli, zero na zero, edytuj, spacja` but after modification: `Tabela, Przejdź do tabeli, 5 na 6, wejście angielski, kolumna 3 z 5, koniec wiersza`. So imo, it's even worse.

Isn't "wejście angielski" supposed to mean that the field is editable and its language is set to English? 🤔 Yet it's super cryptic (at least in Polish).

> I'm not sure. Do you mean adding aria-label to td?

More of `[aria-describedby]` on the `table`/`figure` with the dimensions of the table.

**Comment by Mati365** [comment]
> What screen reader did you test? VoiceOver? It would be better to test it in English and to also test JAWS and NVDA.

Microsoft Reader using this HTML file:
[index.html](https://github.com/user-attachments/files/16930161/index.html.txt)

> Isn't "wejście angielski" supposed to mean that the field is editable and its language is set to English? 🤔 Yet it's super cryptic (at least in Polish).

🤔  Good point. I'll check English version. 

> More of [aria-describedby] on the table/figure with the dimensions of the table.

I'll check that.

**Comment by Mati365** [comment]
Looks like this HTML structure works reasonable fine on Microsoft Narrator:

```html
<td class="ck-editor__editable ck-editor__nested-editable" role="cell" tabindex="-1" aria-label="Edit table cell" contenteditable="true">
```

It doesn't announce `0 by 0`, but the problem is that NVDA / Narrator and JAWS no longer treat field as editable and screen reader navigation seems to be a lot more difficult.

**Comment by Mati365** [comment]
I added draft PoC of table with adjusted HTML structure that works on Microsoft Narrator: https://github.com/ckeditor/ckeditor5/pull/17083

Unfortunately, it looks it's buggy and needs refactor of major parts of table CSS and investigation why Firefox hangs.

**Comment by Witoso** [comment]
@Mati365 could you prepare a comparison table of screen readers of current solution vs the one you tested.

**Comment by Mati365** [comment]
@Witoso Checked few screen readers on that table:

![obraz](https://github.com/user-attachments/assets/3857f823-e906-4926-b460-f9fee5b3d3df)

It looks like changing structure of table HTML improved support for Microsoft Narrator and JAWS. NVDA works the same as before.

I did not test Apple VoiceOver.

### Microsoft Narrator: 

It's improved ✅  `0 by 0` issue has been fixed.

Before: `Enter table, 0 by 0, <content editable content>`
After: `Enter table, 10 by 12, <content editable content>`

### NVDA

Nothing changed, but it was ok already. ✅ 

Before: `Table, Object with Data, Editing fields, <content editable content>`
After: `Table, Object with Data, Editing fields, <content editable content>`

### JAWS

It looks like changing structure of table improved JAWS navigation a lot. ✅ 

Before: `Type and test`
After: `Terrestrial planets, Diameter (km), (from time to time coordinates of cell in the table), Type and test`

**Comment by Reinmar** [comment]
My 2cs about inserting `<div>s` into `<td>s` to move `contenteditable` there. I think it's a no-go solution due to how it increases complexity and risks:

* Those `<div>s` will be prone to styling issues. Margins, paddings, etc, will further affect the current layout. The bogus-paragraphs we use there already cause many issues. 
* Additional elements will have potentially have impact on performance (more to render). Rendering large tables necessitate rendering huge numbers of elements already, and with such change we'd worsen this.

I'd test the other options proposed by @Comandeer and one more to the table (something we were considering with @niegowski many years ago): completely stopping using `contenteditable=true/false` for widgets (or at least tables). We intercept pretty much every interaction and our model/view guards potential selection states anyway.

**Comment by Reinmar** [comment]
One more thing: looking at https://github.com/ckeditor/ckeditor5/issues/16923#issuecomment-2289026967, are we actually sure that there's one solution that would satisfy all these engines? All seem to be have differently now. 

What's the chance that by tuning the table's DOM structure we'll actually make all engines to announce tables correctly, taken for instance that they are a widget?

And finally, can't we simply use the `options.label` option of the `toWidget()` function to describe the element? Won't it in any way affect how the content is announced in this case?

**Comment by Witoso** [comment]
To summarize, we have the following options:
* remove the textbox role – however, it could lead to loss of information about cells being editable; needs testing,
* keep the current structure and provide the table info ourselves, or as @Reinmar put it can't we simply use the options.label option of the toWidget() function to describe the element?

**Comment by Witoso** [comment]
Plan:
1. Prepare PoC for each approach
    1. div (use the previous poc, align with recent master)
    2. remove the textbox role 
    3. additional widget label on a table (combined with textbox role from point 2.)
2. Give it to tests (Narrator, NVDA, VoiceOver, JAWS)
    1. Entering table,
    2. Entering cells,
4. For now, we keep `contenteditable=true/false` as backup.

**Comment by Mati365** [comment]
## Before changes

https://github.com/user-attachments/assets/c352edaf-5271-4cdf-b43f-dfd260958a83

----

## After changes

### `div` with `contenteditable` within `td`.

PR: https://github.com/ckeditor/ckeditor5/pull/17083
POCer: https://pocer.ckeditor.dev/Microsoft%20Narrator%20tables%20PoC%20(extra%20divs%20within%20td)/2025-03-05%2009-37-39/ckeditor5-table/tests/manual/table.html

https://github.com/user-attachments/assets/60054854-f94e-463b-ad97-f980a259aa84

### Remove the textbox role.  

PR: https://github.com/ckeditor/ckeditor5/pull/18052
POCer: https://pocer.ckeditor.dev//Microsoft%20Narrator%20tables%20PoC%20(remove%20role%3Dtextbox)/2025-03-05%2009-48-36

https://github.com/user-attachments/assets/05f8b236-a47a-40f1-aaa4-dbd64d340398

### `toWidget` with option `label`

I used `toWidget()` and `label` option and it is ignored. The default widget label is also ignored (`Press Enter to type after or press Shift + Enter to type before the widget`). It looks like Microsoft Narrator 

PR: https://github.com/ckeditor/ckeditor5/pull/18054
POCer: https://pocer.ckeditor.dev//Microsoft%20Narrator%20tables%20PoC%20(widget%20label)/2025-03-05%2009-56-06

### `aria-describedby` on whole table / `aria-label` on cell

I checked these two attributes. `aria-describedby` is ignored on `table` and `figure` by Microsoft Narrator. `aria-label` is not ignored, thought.

[aria-describedby.html.txt](https://github.com/user-attachments/files/19085305/aria-describedby.html.txt)

**Comment by Witoso** [comment]
@Mati365 please schedule the tests per
> Give it to tests (Narrator, NVDA, VoiceOver, JAWS)
> Entering table,
> Entering cells,

**Comment by Mati365** [comment]
Done

**Comment by Mati365** [comment]
Fyi, I reopened PRs. They were auto-closed after removing `master-it84`.

**Comment by Witoso** [comment]
As removing `role=textbox` doesn't degrade the experience in most screen readers, but rather improves it (with table dimensions appearing in NVDA, Narrator, JAWS) we decided to implement this solution.
